### PR TITLE
ssl/npn+alpn: adjust protocol selection to fail if no protocols match

### PR DESCRIPTION
**Pull request by jmesmon**
ssl/npn+alpn: adjust protocol selection to fail if no protocols match
The current behavior causes a server written using rust-openssl to (if
it cannot negotiate a protocol) fallback to the first protocol it has
avaliable.

This makes it impossible to detect protocol mismatches.

This updates our selection to be more similar to how openssl's
s_server behaves: non-matching protocols are not supplied with a
fallback.

Note that some setups may actually want a fallback protocol supplied
via ALPN. To support those cases, we should consider adding a generic
callback that allows protocol selection to be entirely controlled by
the programmer.

For the purposes of having a sane default, however, not supplying a
default (and mimicing s_server's behavior) is the best choice.

### CODE REVIEW

**Comment by sfackler** [comment]
This seems reasonable. Would it be possible to add a test to make sure it behaves properly?

**Comment by jmesmon** [comment]
Updated to add a test.

File: openssl/src/ssl/tests.rs [hunk 1]
```diff
@@ -628,6 +628,48 @@ fn test_alpn_server_advertise_multiple() {
...
+        Ok(_) => {}
+        Err(err) => panic!("Unexpected error {:?}", err)
+    }
+    // Now connect to the socket and make sure the protocol negotiation works...
+    let stream = TcpStream::connect(localhost).unwrap();
+    let stream = match SslStream::new(&ctx, stream) {
+        Ok(stream) => stream,
+        Err(err) => panic!("Expected success, got {:?}", err)
+    };
+    // SPDY is selected since that's the only thing the client supports.
```
Comment (sfackler) [inline openssl/src/ssl/tests.rs]: Comment's not correct here, right?

Comment (jmesmon) [inline openssl/src/ssl/tests.rs]: whoops. commit updated to fix.

**Pull request merged**